### ISSUE TITLE
Refreshable authentication token

### ISSUE THREAD

**Issue by VakarisZ** [Flask-Middleware/flask-security#774]
The token that `flask-security-too` provides when `?include_auth_token` query parameter is passed is not refreshable. This means that when a token expires, a new token must be generated through a login process. 

We're interested in a feature that would allow us to refresh the tokens that are actively used but expire the ones that are not. My idea so far is to store `last_access` time date field in the user model and update it on each successful token usage. When validating the token, the `max_age` would be extended based on the `last_access` field of the user. This would ensure that active tokens are not getting expired.

Is there a way to refresh the token without forcing the client to re-login? Does my proposed solution sound like something you would be willing to accept as a PR? Thanks!

**Comment by jwag956** [comment]
Well - in some ways they are refreshable - if you do a GET /login?included_auth_token, you will get a new token - not quite as 'automatic' as your proposal (which I need to think a bit more about). At first glance - I am not keen on having users pass in expired tokens and accepting them because of other out-of-band data.. (that's not how for example oauth refresh tokens work).

**Comment by VakarisZ** [comment]
For our case the re-login is more complicated as we have a custom authentication scheme. When I think of it, you're right - the major flaw I overlooked is that a user might have multiple tokens active at a time (for our use case this won't be happening), thus we must save data in the database for each token in this case. 

What about an alternative approach, a refresh token with an extra lifespan that can be used when the first one expires to fetch a new one?

**Comment by jwag956** [comment]
You don't have to 're-login' just hit the /login endpoint with a GET - no need for username/password or anything, use ?include_auth_token and you will get a new fresh token.
We should probably not re-invent oauth2 :-) however having the client code store 2 tokens doesn't really help anything - if the client is compromised, having the refresh token allows them to get a new token (in oauth2 the client does NOT get the refresh token).

**Comment by VakarisZ** [comment]
> You don't have to 're-login' just hit the /login endpoint with a GET - no need for username/password or anything, use ?include_auth_token and you will get a new fresh token.

So the client is expected to periodically refresh the token? This is not a good approach, since now the client has to have an asynchronous process just to make sure the token it uses does not expire. Also, the expiration date is not in the token, so the client does not know how often he needs to query this endpoint to refresh. A better workflow: the client gets a refresh token and if the auth token stops working the refresh token can be used to get a new pair. 

> We should probably not re-invent oauth2 :-)

So am I expected to set up a custom authorization server just to have refreshable tokens?

> however having the client code store 2 tokens doesn't really help anything - if the client is compromised, having the refresh token allows them to get a new token

Refresh tokens usually have an expiration time, so it helps if a stale client gets compromised.

> in oauth2 the client does NOT get the refresh token

[Based on this](https://oauth.net/2/refresh-tokens/) looks like it does.

**Comment by jwag956** [comment]
So here is a quick idea....
- override User.get_security_payload() - and for requests that have the ?include_auth_token, use current_user.get_auth_token() to generate another (2nd) auth_token. This will return 2 tokens to the client. See: https://flask-security-too.readthedocs.io/en/stable/models.html#custom-user-payload
-  add an endpoint in your application that takes this 2nd token (e.g. /refresh/{token} and creates a new token.

Now - the tricky part is that the POST /refresh/{token} (and it should be a POST) has to:
-  parse the token (using flask_security.check_and_get_token_status. The serializer_name should be "remember".)
-  that will return a tuple (expired, invalid, data) - assuming it is valid then
- look up the user with app.security.datastore.find_user(fs_uniquifier=data[0]) (or if you've defined fs_token_uniquifier use that).
- It that returns a valid user - use user.get_auth_token() and returns that

**Comment by VakarisZ** [comment]
Thanks for the help. The authentication token used in flask-security-too is basically a JWT and JWT's offer refreshability: https://flask-jwt-extended.readthedocs.io/en/stable/refreshing_tokens/#refreshing-tokens . 

The main reason we opened an issue is to solve the issue of non-refreshable tokens for this package (since we have to do it for ourselves we might as well contribute). Would you be willing to accept a PR that enables refreshability with lets say `REFRESHABLE_TOKENS` and also uses `REFRESH_TOKEN_EXPIRES` config value to set the lifetime of a refresh token?  This doesn't seem like a complex feature to implement, since we can use the exact same mechanisms of token generation and validation, except refresh tokens will have a longer lifespan.

**Comment by VakarisZ** [comment]
Hey @jwag956, have you thought about adding the refresh token to the package? We can override the methods, but we rather contribute to make the `flask-security-too` configurable with refreshable tokens and refresh token lifespan. Would you be willing to accept a PR like that or is it out of consideration?

**Comment by jwag956** [comment]
The TL;DR is that I don't think this is a widely applicable feature.

I don't know anything about your application, application architecture nor what security posture you are trying to achieve - so don't take this as me trying to convince you not to go down this route.

Here is my thinking. First and foremost - Flask-Security (FS) auth-tokens are NOT bearer tokens - so one reason to have short-lived access-tokens is because in oauth and many JWT implementations they ARE bearer tokens and therefore there is no easy way to invalidate them (neither the user nor an administrator). But FS auth-tokens can easily be revoked by the user or administrator (this also means that FS isn't near as scalable since there is a DB lookup on each API request).

Second, in most cases (see below), when using oauth2, the 'client' isn't the 'user' - the client is the web-app - and refresh tokens aren't actually sent to the end-user (web browser). Furthermore in most implementations, a refresh-token isn't a bearer token - in order to fetch a new access-token the 'client' requires the users refresh-token and its  own client ID.

FS's auth-tokens are more closely aligned with oauth2s 'client credential grant' - which does in fact require the client to periodically re-authenticate (usually with a client ID /client secret pair stored in the environment).

There is one use case for refresh tokens in the browser - so-called SPA applications - here is a link for that:
https://www.pingidentity.com/en/resources/blog/post/refresh-token-rotation-spa.html

note that to implement that securely they talk about refresh token rotation.

It is also important to consider that FS isn't trying to implement oauth2 - if that's the level of authentication you need - auth0 is a great package that has a huge amount of boilerplate readily available.

Lastly, FS strongly believes in using sessions for any web-based application - if you need more security, use server-side sessions.

**Comment by VakarisZ** [comment]
> FS's auth-tokens are more closely aligned with oauth2s 'client credential grant' - which does in fact require the client to periodically re-authenticate (usually with a client ID /client secret pair stored in the environment).

So why can I fetch a new token without re-authentication (If I have an already valid token, I can just query the login endpoint and fetch a new token with a longer lifespan)?

> auth0 is a great package that has a huge amount of boilerplate readily available.

It's an overkill for us. We need to be able to revoke credentials (auth0 uses JWT's), we don't need third party authorization (sign up through facebook for example), we need roles and decorators that FS offers, etc.

> FS strongly believes in using sessions for any web-based application - if you need more security, use server-side sessions

We have 2 clients: SPA and an API client. None of them use cookies. It seems like sessions are tightly related to setting the cookies. I also don't see a mechanism to invalidate stale sessions by default.

Bottom line is that either FS shouldn't allow you to fetch the token without credentials or it should have a proper refresh mechanism. Sending a token to get a new token sucks as a refresh mechanism for the reasons mentioned before (the client doesn't know when it needs to refresh the token, the client needs to keep track of time for the token refresh, etc.)

**Comment by jwag956** [comment]
So looking back at the code - I was completely wrong about being able to get a new auth_token - the only way you can get an auth token is the result of a POST with credentials.

sessions are tied to cookies and are implemented in Flask itself. For SPA I would highly recommend using cookies - easy, secure (especially with same-site) small (smaller if you use server side cookies). sessions can be browser session based or timed. With FS you can revoke all sessions if an account becomes compromised.

... more to add here - will get to that in the next day or so.

**Comment by VakarisZ** [comment]
> So looking back at the code - I was completely wrong about being able to get a new auth_token

I somehow also presumed that I've seen this behavior, but I've tested and couldn't reproduce it 🤔 

I guess we'll have to override the library to generate a separate refresh token with a bit longer lifespan. If you're interested in adding this to the FS let us know and we'll open up a PR. Thank you for all your help!
